core/src/main/java/org/elasticsearch/ElasticsearchException.java

```diff
@@ -324,18 +335,72 @@ public static void toXContent(XContentBuilder builder, Params params, Throwable
         if (ex instanceof ElasticsearchException) {
             ((ElasticsearchException) ex).toXContent(builder, params);
         } else {
-            builder.field("type", getExceptionName(ex));
-            builder.field("reason", ex.getMessage());
+            builder.field(TYPE, getExceptionName(ex));
+            builder.field(REASON, ex.getMessage());
             if (ex.getCause() != null) {
-                builder.field("caused_by");
+                builder.field(CAUSED_BY);
                 builder.startObject();
                 toXContent(builder, params, ex.getCause());
                 builder.endObject();
             }
             if (params.paramAsBoolean(REST_EXCEPTION_SKIP_STACK_TRACE, REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT) == false) {
-                builder.field("stack_trace", ExceptionsHelper.stackTrace(ex));
+                builder.field(STACK_TRACE, ExceptionsHelper.stackTrace(ex));
+            }
+        }
+    }
+
+    /**
+     * Generate a {@link ElasticsearchException} from a {@link XContentParser}. This does not
+     * return the original exception type (ie NodeClosedException for example) but just wraps
+     * the type, the reason and the cause of the exception. It also recursively parses the
+     * tree structure of the cause, returning it as a tree structure of {@link ElasticsearchException}
+     * instances.
+     */
+    public static ElasticsearchException fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token = ensureFieldName(parser, parser.nextToken());
+
+        String type = null, reason = null, stack = null;
+        ElasticsearchException cause = null;
+        Map<String, Object> headers = new HashMap<>();
+
+        do {
+            String currentFieldName = parser.currentName();
+            token = parser.nextToken();
+            if (token.isValue()) {
+                if (TYPE.equals(currentFieldName)) {
+                    type = parser.text();
+                } else if (REASON.equals(currentFieldName)) {
+                    reason = parser.text();
+                } else if (STACK_TRACE.equals(currentFieldName)) {
+                    stack = parser.text();
+                } else {
+                    // Everything else is considered as a header
+                    headers.put(currentFieldName, parser.text());
+                }
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (CAUSED_BY.equals(currentFieldName)) {
+                    cause = fromXContent(parser);
+                } else if (HEADER.equals(currentFieldName)) {
+                    headers.putAll(parser.map());
+                } else {
+                    throwUnknownField(currentFieldName, parser.getTokenLocation());
+                }
             }
+        } while ((token = parser.nextToken()) == XContentParser.Token.FIELD_NAME);
```

Comment:
Yes, I think this is true for additional metadata of type objects - other fields/value should be reparsed as header by

```
// Everything else is considered as a header
headers.put(currentFieldName, parser.text());
```

I'll add a test for this case, thanks @Mpdreamz for pointing that!